src/partial.rs

```diff
@@ -843,6 +846,220 @@ impl DisplayWithPartialPaths for PartialScopeStack {
+    /// `None`.
+    pub fn pop_front(&mut self, partials: &mut PartialPaths) -> Option<PartialPathEdge> {
+        let result = self.edges.pop_front(&mut partials.partial_path_edges);
+        if result.is_some() {
+            self.length -= 1;
+        }
+        result.copied()
+    }
+
+    /// Removes and returns the edge at the back of this edge list.  If the list is empty, returns
+    /// `None`.
+    pub fn pop_back(&mut self, partials: &mut PartialPaths) -> Option<PartialPathEdge> {
+        let result = self.edges.pop_back(&mut partials.partial_path_edges);
+        if result.is_some() {
+            self.length -= 1;
+        }
+        result.copied()
+    }
+
+    pub fn display<'a>(
+        self,
+        graph: &'a StackGraph,
+        partials: &'a mut PartialPaths,
+    ) -> impl Display + 'a {
+        display_with(self, graph, partials)
+    }
+
+    /// Returns whether one edge list shadows another.  Note that shadowing is not commutative — if
+    /// path A shadows path B, the reverse is not true.
+    pub fn shadows(mut self, partials: &mut PartialPaths, mut other: PartialPathEdgeList) -> bool {
+        while let Some(self_symbol) = self.pop_front(partials) {
+            if let Some(other_symbol) = other.pop_front(partials) {
+                if self_symbol.shadows(other_symbol) {
+                    return true;
+                }
+            } else {
+                return false;
+            }
+        }
+        false
+    }
+
+    pub fn equals(mut self, partials: &mut PartialPaths, mut other: PartialPathEdgeList) -> bool {
+        while let Some(self_symbol) = self.pop_front(partials) {
+            if let Some(other_symbol) = other.pop_front(partials) {
+                if self_symbol != other_symbol {
+                    return false;
+                }
+            } else {
+                return false;
+            }
+        }
+        other.edges.is_empty()
+    }
+
+    pub fn cmp(
+        mut self,
+        partials: &mut PartialPaths,
+        mut other: PartialPathEdgeList,
+    ) -> std::cmp::Ordering {
+        use std::cmp::Ordering;
+        while let Some(self_symbol) = self.pop_front(partials) {
+            if let Some(other_symbol) = other.pop_front(partials) {
+                match self_symbol.cmp(&other_symbol) {
+                    Ordering::Equal => continue,
+                    result @ _ => return result,
+                }
+            } else {
+                return Ordering::Greater;
+            }
+        }
+        if other.edges.is_empty() {
+            Ordering::Equal
+        } else {
+            Ordering::Less
+        }
+    }
+
+    /// Returns an iterator over the contents of this edge list.
+    pub fn iter<'a>(
+        &self,
+        partials: &'a mut PartialPaths,
+    ) -> impl Iterator<Item = PartialPathEdge> + 'a {
+        self.edges.iter(&mut partials.partial_path_edges).copied()
```

Comment:
The short version is that if `PartialPathEdge` were a large struct, I'd go with exactly what you ended up with at the end.  Return a _reference_ to the elements of the list so that you don't have to spend the cycles copying the contents of each element.  And that requires the lifetime annotations you came up with so that the compiler can enforce that the list can't go away while you're doing stuff with the element references that you get.

In this case, though, `PartialPathEdge` is just two `u32`s.  On 64-bit platforms that means that the struct itself is the same size as the pointers that you'd get by iterating over the elements by reference.  So it's actually cheaper to iterate by value!  `.copied()` is just the idiomatic way to do that since `PartialPathEdge` implements `Copy`.  It's equivalent to `.map(|x| *x)`.

Reply:
@dcreager that's very helpful, thank you for the explanation!